// feathers-authentication/lib/client/utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var getCookie = exports.getCookie = function getCookie(name) {
  var value = '; ' + document.cookie;
  var parts = value.split('; ' + name + '=');

  if (parts.length === 2) {
    return parts.pop().split(';').shift();
  }

  return null;
};

var getUser = exports.getUser = function getUser() {
  // TODO (EK): Maybe make this configurable
  var key = 'feathers-user';
  var user = localStorage.getItem(key);

  return JSON.parse(user);
};

var setToken = exports.setToken = function setToken(token) {
  // TODO (EK): Maybe make this configurable
  var key = 'feathers-jwt';
  localStorage.setItem(key, token);

  // TODO (EK): Support async storage for react native 

  return true;
};

var setUser = exports.setUser = function setUser(user) {
  // TODO (EK): Maybe make this configurable
  var key = 'feathers-user';
  localStorage.setItem(key, JSON.stringify(user));

  // TODO (EK): Support async storage for react native 

  return true;
};

var getToken = exports.getToken = function getToken() {
  // TODO (EK): Maybe make this configurable
  var key = 'feathers-jwt';
  var token = localStorage.getItem(key);

  if (token) {
    return token;
  }

  // TODO (EK): Support async storage for react native 

  // We don't have the token so try and fetch it from the cookie
  // and store it in local storage.
  // TODO (EK): Maybe we should clear the cookie
  token = getCookie(key);

  if (token) {
    localStorage.setItem(key, token);
  }

  return token;
};

var clearToken = exports.clearToken = function clearToken() {
  // TODO (EK): Maybe make this configurable
  var key = 'feathers-jwt';

  // TODO (EK): Support async storage for react native
  localStorage.removeItem(key);

  return true;
};

var clearUser = exports.clearUser = function clearUser() {
  // TODO (EK): Maybe make this configurable
  var key = 'feathers-user';

  // TODO (EK): Support async storage for react native
  localStorage.removeItem(key);

  return true;
};

exports.default = {
  getUser: getUser,
  setUser: setUser,
  clearUser: clearUser,
  getToken: getToken,
  setToken: setToken,
  clearToken: clearToken,
  getCookie: getCookie
};
